### ISSUE TITLE
Iterated padding fails

### ISSUE THREAD

**Issue by philipp-harms** [beancount/beancount#719]
In two consecutive pad commands, the second one doesn't work.

```
option "operating_currency" "USD"

2020-01-01 commodity USD
2022-01-01 open Assets:A
2022-01-01 open Assets:B
2022-01-01 open Assets:C

; transfer 50 USD from A to B 
2022-01-03 pad Assets:A Assets:B
2022-01-04 balance Assets:A -50 USD
; as expected, there are now -50 USD in Assets:A and 50 USD in Assets:B

; transfer 50 USD from B to C
2022-01-05 pad Assets:B Assets:C
2022-01-06 balance Assets:B 0 USD
; this pad command has no effect, and the balance assertion fails
```

**Comment by blais** [comment]
Looks like an important bug indeed.

**Comment by blais** [comment]
That will take some work, is a known issue (padding is implemented per target account, in isolation). Essentially must rewrite the padding algorithm in its entirety, and it's not trivial, since a degenerate case of your example can result in a chain of mutually affecting padding directives across many accounts.

I think we have to rewrite the algorithm to perform across all accounts, likely sorting by next-balance-date instead of per account. I have to think about this.

**Comment by philipp-harms** [comment]
Thanks for the quick answer!

> degenerate case of your example 

I use such pad commands regularly when restructuring my account hierarchy. For instance, when deleting an expense or liability sub-account, I want to move all of what's in account A into some other account B. Transactions are less flexible than paddings because the transaction amount would have to be changed manually whenever a past expense is re-categorized. 

Sometimes, I would prefer to to move only a fraction into the new account, but this seems impossible without plugins. My preferred solution would be to write transactions with variable amounts (e.g. arithmetic expressions involving account balances).

**Comment by blais** [comment]
Actually, I can prove by contradiction that a solution that would work across accounts is ill-defined.
Consider this; two accounts, 1 and 2.
Account 1 has initial balance A1 of 100, then a padding P1 on day 1 against account 2 and a balance B1 on day 3 of 150.
Account 2 has initial balance A2 of 110, then a padding P2 on day 2 against account 1 and a balance B2 on day 3 of 150 also.
You can show that to balance we need to solve for values of P1 and P2 which satisfy:

B1 = A1 + P1 - P2
B2 = A2 - P1 + P2

Extend it to a larger system of equations...

**Comment by philipp-harms** [comment]
True, but your example neglects the [recommendation](https://beancount.github.io/docs/beancount_language_syntax.html#pad) that the first account in a pad directive should have a balance assertion (immediately) following it. In contrast, my example adheres to this recommendation and therefore does not require any fancy cross-account computation. 

My suggestion is to process the pad directives in ascending temporal order and to ignore all future pad directives when computing the padding. This works whenever the above recommendation is followed and throws a failed-balance error otherwise. For instance, my example would be processed as expected with P1=P2=50, whereas your example would resolve to P1=50 and P2=200, resulting in a failed balance for account 1 and a correct one for account 2.

**Comment by philipp-harms** [comment]
The cross-account computation isn't that difficult after all. 
```
m = [number of balance statements]
n = [number of pad statements]
B = [monetary vector of size m]
for i = 1 to m do
  B(i) = [i-th balance]
  B(i) = B(i) - [balance preceding the i-th balance on the same account]
  B(i) = B(i) - [sum of all postings between the two balances, ignoring pad commands]
end for
M = [integer matrix of size m times n, initialized to zero] 
for j = 1 to n do
  s = [index of the balance statement on pad j's source account succeeding pad j]
  t = [index of the balance statement on pad j's target account succeeding pad j]
  M(s,j) = -1
  M(t,j) = 1
end for  
P = [vector of size n which solves M*P=B]
```
You'll get information about existence and uniqueness of solutions from the linear algebra package.
